crates/transpiler/src/passes/disjoint_layout.rs

```diff
@@ -0,0 +1,545 @@
+use rustworkx_core::petgraph::EdgeType;
+use smallvec::SmallVec;
+use uuid::Uuid;
+
+use qiskit_circuit::bit::ShareableQubit;
+use qiskit_circuit::converters::circuit_to_dag;
+use qiskit_circuit::dag_circuit::DAGCircuit;
+use qiskit_circuit::imports::ImportOnceCell;
+use qiskit_circuit::operations::{Operation, OperationRef, Param, StandardInstruction};
+use qiskit_circuit::packed_instruction::PackedOperation;
+use qiskit_circuit::{Clbit, PhysicalQubit, Qubit, VirtualQubit};
+
+use crate::target::{Qargs, Target};
+use crate::TranspilerError;
+
+create_exception!(qiskit, MultiQEncountered, pyo3::exceptions::PyException);
+
+static COUPLING_MAP: ImportOnceCell =
+    ImportOnceCell::new("qiskit.transpiler.coupling", "CouplingMap");
+
+pub struct DisjointComponent {
+    pub physical_qubits: Vec<PhysicalQubit>,
+    pub sub_dag: DAGCircuit,
+    pub virtual_qubits: Vec<VirtualQubit>,
+}
+
+pub enum DisjointSplit {
+    NoneNeeded,
+    TargetSubset(Vec<PhysicalQubit>),
+    Arbitrary(Vec<DisjointComponent>),
+}
+
+type CouplingMap = StableDiGraph<NodeIndex, ()>;
+
+fn subgraph(graph: &CouplingMap, node_set: &HashSet<NodeIndex>) -> CouplingMap {
+    let mut node_map: HashMap<NodeIndex, NodeIndex> = HashMap::with_capacity(node_set.len());
+    let node_filter = |node: NodeIndex| -> bool { node_set.contains(&node) };
+    let mut out_graph = CouplingMap::with_capacity(node_set.len(), node_set.len());
+    let filtered = NodeFiltered(&graph, node_filter);
+    for node in filtered.node_references() {
+        let new_node = out_graph.add_node(*node.1);
+        node_map.insert(node.0, new_node);
+    }
+    for edge in filtered.edge_references() {
+        let new_source = *node_map.get(&edge.source()).unwrap();
+        let new_target = *node_map.get(&edge.target()).unwrap();
+        out_graph.add_edge(new_source, new_target, ());
+    }
+    out_graph
+}
+
+#[pyfunction(name = "run_pass_over_connected_components")]
+pub fn py_run_pass_over_connected_components(
+    dag: &mut DAGCircuit,
+    target: &Target,
+    run_func: Bound<PyAny>,
+) -> PyResult<Option<Vec<PyObject>>> {
+    let func = |dag: DAGCircuit, cmap: &CouplingMap| -> PyResult<PyObject> {
+        let py = run_func.py();
+        let coupling_map_cls = COUPLING_MAP.get_bound(py);
+        let endpoints: Vec<[usize; 2]> = cmap
+            .edge_indices()
+            .map(|edge| {
+                let endpoints = cmap.edge_endpoints(edge).unwrap();
+                [endpoints.0.index(), endpoints.1.index()]
+            })
+            .collect();
+        let out_list = PyList::new(py, endpoints)?;
+        let py_cmap = coupling_map_cls.call1((out_list,))?;
+        for node in cmap.node_indices() {
+            py_cmap
+                .getattr(intern!(py, "graph"))?
+                .set_item(node.index(), cmap.node_weight(node).unwrap().index())?;
+        }
+        // TODO: Remove the dag clone
```

Comment:
I think this comment is at least misplaced now?

Reply:
Removed in: https://github.com/Qiskit/qiskit/pull/14177/commits/5fdcf92a63fbe5a64456c85de4a3ac9722a6dedd